feat: add links to the author bio

laurelfulford opened the pull request:
### All Submissions:

* [x] Have you followed the [Newspack Contributing guideline](https://github.com/Automattic/newspack-theme/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) and [VIP Go coding standards](https://vip.wordpress.com/documentation/vip-go/code-review-blockers-warnings-notices/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:

This PR adds links from the author bio avatar and author name to each author's archive, and makes a small change to the sponsor bio to make the link behaviour similar. 

It also fixes a small bug with the Newspack default theme's author bio link contrast, likely introduced in #2091. Rather than switching to a medium grey when the primary colour is too like, the 'More by...' link continued to use the too-light primary colour. 

See 1200550061930446-as-1204508674733864.

### How to test the changes in this Pull Request:

**Test the link updates for authors:**
1. Apply the PR and run `npm run build`
2. Create a post that has both a regular and Co-Author Plus guest author assigned to it. 
3. Confirm that each author's avatar and name link to their archive page.
4. Turn off Co-Author Plus and confirm it still works with the regular author. 
5. Switch to Newspack Sacha, and repeat steps 2 and 3 (this theme has a different template part for the author bio to achieve the centred layout). 

**Test the link update for sponsors:**
This change involved adding the "Sponsored by" text into the existing sponsor link in the bio, just to make the click area more consistent with regular authors (whose whole title is clickable).

1. Assign a Sponsor to the post, and make sure they have a link
2. Confirm their whole title name "Sponsored by [Name]", not just the name part, links to their website in the sponsor bio at the bottom of the post. 
3. Remove the URL from the Sponsor's settings and confirm nothing weird happens to the sponsor output. 

**Test the link contrast:**
1. Switch to the Newspack default theme
2. Switch the Primary colour to a dark colour and confirm it's being used for the "More by..." link at the end of the bio.
3. Switch the Primary colour to a light colour and confirm the "More by..." link falls back to a medium grey that's still legible against the white background. 

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [ ] Have you successfully ran tests with your changes locally?

<!-- Mark completed items with an [x] -->

miguelpeixe reviewed (changes requested):
While testing the sponsor I got a spacing when the sponsor doesn't have a logo:

| With sponsor logo | Without sponsor logo |
| --- | --- |
| <img width="332" alt="image" src="https://github.com/Automattic/newspack-theme/assets/820752/a2a70f15-bd5c-4ead-826d-d043eedc8518"> | <img width="300" alt="image" src="https://github.com/Automattic/newspack-theme/assets/820752/86c9a2d8-d7ce-4f8a-a464-3b0be1b4c714"> |

> Switch the Primary colour to a dark colour and confirm it's being used for the "More by..." link at the end of the bio.

Not sure if expected, but on Sacha I get the same medium grey that is applied for the light primary color. It works as expected for the default theme.

leogermani commented:
When I switched to this branch I was getting a Fatal error:

```
PHP Parse error:  syntax error, unexpected token ":" in /newspack-repos/newspack-theme/newspack-theme/template-parts/post/author-bio.php on line 86
```

And it was because of [this PHP short opening tag](https://github.com/Automattic/newspack-theme/pull/2114/commits/b7be5adf74540a044b57cb63fd94d1c026c364ad#diff-2b452ca618c9c3fbe5e304894273db7ebfbf45a8bc778f6b8950e3a97eabb447R32). Some environments won't consider those (including mine) and that's why PHP linter was complaining about it. The closing `}` was not being processed.

I fixed that in https://github.com/Automattic/newspack-theme/pull/2114/commits/fb3b9caf9b146b07d190437c4741b57019a460da but also did some changes to make it more readable in https://github.com/Automattic/newspack-theme/pull/2114/commits/7b029e2d2edc74b22fab241bc81a88c453f401a9

I wanted to fix more things but the indentation in that file is a mess... so I just made some small tweaks.

laurelfulford commented:
> While testing the sponsor I got a spacing when the sponsor doesn't have a logo:

Thanks @miguelpeixe! I didn't touch the byline space in this PR, just the author bio at the bottom, so I added a fix for this in a separate PR (#2120). 

> Not sure if expected, but on Sacha I get the same medium grey that is applied for the light primary color. It works as expected for the default theme.

That is correct, that's my bad! 🤦‍♀️ It should be using the primary colour for just the default theme, and the secondary colour for the rest! (At least, it's correct if the secondary colour you're using is lighter!). I double-checked against master to make sure. 

> I fixed that in https://github.com/Automattic/newspack-theme/commit/fb3b9caf9b146b07d190437c4741b57019a460da but also did some changes to make it more readable in https://github.com/Automattic/newspack-theme/commit/7b029e2d2edc74b22fab241bc81a88c453f401a9

Thanks @leogermani!
